app/photosketchingcontroller.cpp

```diff
@@ -0,0 +1,204 @@
+#include <QUrl>
+
+#include "coreutils.h"
+
+/**
+ * ColorPath declarations
+ */
+
+
+ColorPath::ColorPath()
+{
+  mColor = QColor();
+  mPoints = QVector<QPointF>();
+}
+
+bool ColorPath::operator==( const ColorPath &other ) const
+{
+  if ( mColor == other.mColor  &&  mPoints == other.mPoints )
+  {
+    return true;
+  }
+
+  return false;
+}
+
+
+/**
+ * PhotoDrawingController declarations
+ */
+
+
+void PhotoSketchingController::newSketch()
+{
+  mCurrentLine = ColorPath( mPenColor, {} );
+}
+
+void PhotoSketchingController::addPoint( const QPointF newPoint )
+{
+// we scale up the point to picture's true position
+  const QPointF realPoint = QPointF( newPoint.x() * mPhotoScaleRatio, newPoint.y() * mPhotoScaleRatio );
+  mCurrentLine.mPoints.append( realPoint );
+  if ( mPaths.isEmpty() || mCurrentLine.mPoints.size() == 1 )
+  {
+    mPaths.append( mCurrentLine );
+    emit newPathAdded( -1 );
+    if ( !mCanUndo && !mPaths.isEmpty() )
+    {
+      mCanUndo = true;
+      emit canUndoChanged();
+    }
+  }
+  else
+  {
+    mPaths[mPaths.size() - 1] = mCurrentLine;
+    emit pathUpdated( {static_cast<int>( mPaths.size() - 1 ) } );
+  }
+}
+
+void PhotoSketchingController::setActiveColor( const QColor newColor )
+{
+  mPenColor = newColor;
+  mCurrentLine.mColor = newColor;
+
+  emit activeColorChanged();
+}
+
+void PhotoSketchingController::undo()
+{
+  if ( !mPaths.empty() )
+  {
+    mPaths.remove( mPaths.size() - 1 );
+    emit lastPathRemoved();
+  }
+
+  if ( !mPaths.empty() && !mCanUndo )
+  {
+    mCanUndo = true;
+    emit canUndoChanged();
+  }
+  else if ( mPaths.empty() && mCanUndo )
+  {
+    mCanUndo = false;
+    emit canUndoChanged();
+  }
+}
+
+void PhotoSketchingController::clear()
+{
+  mPenColor = QColor::fromString( "#FFFFFF" );
+  emit activeColorChanged();
+  newSketch();
+  mCanUndo = false;
+  emit canUndoChanged();
+  mPaths.clear();
+  emit pathsReset();
+}
+
+void PhotoSketchingController::saveDrawings() const
+{
+  const QString photoPath = QUrl( mPhotoSource ).toLocalFile();
+  QImage image( photoPath );
+  if ( image.isNull() )
+  {
+    CoreUtils::log( "Photo annotations", "Failed to load image from: " + photoPath );
+    return;
+  }
+
+  QPainter painter( &image );
+  painter.setRenderHint( QPainter::Antialiasing );
+
+  const int pathCount = static_cast<int>( mPaths.size() );
+
+  for ( int i = 0; i < pathCount; ++i )
+  {
+    QColor color = mPaths.at( i ).mColor;
+    QVector<QPointF> points = mPaths.at( i ).mPoints;
+
+    if ( points.isEmpty() )
+      continue;
+
+    // if you are adjusting width here don't forget to adjust it also in MMFormPhotoDrawingPageDialog shapePathComponent
+    QPen pen( color, 4 * mPhotoScaleRatio );
+    painter.setPen( pen );
+
+    QPainterPath path;
+    path.moveTo( points.first().x(),  points.first().y() );
+    for ( int j = 1; j < points.size(); ++j )
+    {
+      path.lineTo( points.at( j ).x(),  points.at( j ).y() );
+    }
+
+    painter.drawPath( path );
+  }
```

Comment:
At first I thought this would be more readable like:
```suggestion
  for ( const ColorPath &path : std::as_const( mPaths ) )
  {
    if ( path.mPoints.isEmpty() )
      continue;

    const QPen pen( path.mColor, 4 * mPhotoScaleRatio );
    painter.setPen( pen );
    QPainterPath painterPath;

    bool first = true;
    for ( const QPointF &point : std::as_const( path.mPoints ) )
    {
      if ( first )
      {
        painterPath.moveTo( point );
        first = false;
      }
      else
      {
        painterPath.lineTo( point );
      }
    }
    painter.drawPath( painterPath );
  }
```
but now that I see it again it's not much different!
Maybe just use the `lineTo`/`moveTo` variants accepting `QPointF` instead of `x,y`?

Reply:
Resolved with using `QPointF` variants